projection/validator_delegation/view/delegations.go

```diff
@@ -37,23 +38,151 @@ func NewDelegationsView(handle *rdb.Handle) Delegations {
 	}
 }
 
-func (view *DelegationsView) Clone(previousHeight, currentHeight int64) error {
+func (view *DelegationsView) Insert(row DelegationRow) error {
 
-	return nil
-}
+	sql, sqlArgs, err := view.rdb.StmtBuilder.
+		Insert(
+			"view_vd_delegations",
+		).
+		Columns(
+			"height",
+			"validator_address",
+			"delegator_address",
+			"shares",
+		).
+		Values(
+			fmt.Sprintf("[%v,)", row.Height),
+			row.ValidatorAddress,
+			row.DelegatorAddress,
+			row.Shares.String(),
+		).
+		ToSql()
 
-func (view *DelegationsView) Insert(row DelegationRow) error {
+	if err != nil {
+		return fmt.Errorf("error building delegation insertion sql: %v: %w", err, rdb.ErrBuildSQLStmt)
+	}
+
+	result, err := view.rdb.Exec(sql, sqlArgs...)
+	if err != nil {
+		return fmt.Errorf("error inserting delegation into the table: %v: %w", err, rdb.ErrWrite)
+	}
+	if result.RowsAffected() != 1 {
+		return fmt.Errorf("error inserting delegation into the table: rows inserted: %v: %w", result.RowsAffected(), rdb.ErrWrite)
+	}
 
 	return nil
 }
 
 func (view *DelegationsView) Update(row DelegationRow) error {
 
+	// Check if there is an record's lower bound start with this height.
+	found, err := view.checkIfDelegationRecordExistByHeightLowerBound(row)
+	if err != nil {
+		return fmt.Errorf("error in checking new delegation record existence at this height: %v", err)
+	}
+
+	if found {
+		// If there is a record that height = `[row.Height,)`, then update the existed one
+
+		sql, sqlArgs, err := view.rdb.StmtBuilder.
+			Update(
+				"view_vd_delegations",
+			).
+			SetMap(map[string]interface{}{
+				"shares": row.Shares.String(),
+			}).
+			Where(
+				"validator_address = ? AND delegator_address = ? AND height @> ?::int8",
+				row.ValidatorAddress,
+				row.DelegatorAddress,
+				row.Height,
+			).
+			ToSql()
+
+		if err != nil {
+			return fmt.Errorf("error building delegation update sql: %v: %w", err, rdb.ErrBuildSQLStmt)
+		}
+
+		result, err := view.rdb.Exec(sql, sqlArgs...)
+		if err != nil {
+			return fmt.Errorf("error updating delegation into the table: %v: %w", err, rdb.ErrWrite)
+		}
+		if result.RowsAffected() != 1 {
+			return fmt.Errorf("error updating delegation into the table: row updated: %v: %w", result.RowsAffected(), rdb.ErrWrite)
+		}
+
+	} else {
+		// Else, update the previous record's height range, then insert a new record
+
+		err := view.setDelegationRecordHeightUpperBound(row)
+		if err != nil {
+			return fmt.Errorf("error updating delegation.Height upper bound: %v", err)
+		}
+		err = view.Insert(row)
+		if err != nil {
+			return fmt.Errorf("error inserting a new record for this delegation: %v", err)
+		}
+
+	}
+
 	return nil
 }
 
+func (view *DelegationsView) checkIfDelegationRecordExistByHeightLowerBound(row DelegationRow) (bool, error) {
```

Comment:
suggest `IsExistedByLowerBoundHeight(validatorAddress, delegatorAddress string, height int64) (bool, error)` for readability

Reply:
Thanks! Updated.